src/powershell/Program.cs

```diff
@@ -19,7 +59,439 @@ public sealed class ManagedPSEntry
+                }
+
+                // Check for "-Login" or some prefix thereof
+                if (IsParam(arg, "login", "LOGIN"))
+                {
+                    loginIndex = i;
+                    return true;
+                }
+
+                // After -File and -Command, all parameters are passed
+                // to the invoked file or command, so we can stop looking.
+                if (IsParam(arg, "file", "FILE")
+                    || IsParam(arg, "command", "COMMAND"))
+                {
+                    return false;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Determines if a given parameter is the one we're looking for.
+        /// Assumes any prefix determines that parameter (true for -l, -c and -f).
+        /// </summary>
+        /// <param name="arg">The argument to check.</param>
+        /// <param name="paramToCheck">The lowercase name of the parameter to check.</param>
+        /// <param name="paramToCheckUpper">The uppercase name of the parameter to check.</param>
+        /// <returns></returns>
+        private static bool IsParam(
+            string arg,
+            string paramToCheck,
+            string paramToCheckUpper)
+        {
+            // Quick fail if the argument is longer than the parameter
+            if (arg.Length > paramToCheck.Length + 1)
+            {
+                return false;
+            }
+
+            // Check arg chars in order and allow prefixes
+            for (int i = 1; i < arg.Length; i++)
+            {
+                if (arg[i] != paramToCheck[i-1]
+                    && arg[i] != paramToCheckUpper[i-1])
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Create the exec call to /bin/{z}sh -l -c 'exec -a +pwsh pwsh "$@"' and run it.
+        /// </summary>
+        /// <param name="args">The argument vector passed to pwsh.</param>
+        /// <param name="loginArgIndex">The index of -Login in the argument vector.</param>
+        /// <param name="isMacOS">True if we are running on macOS.</param>
+        /// <param name="pwshPath">Absolute path to the pwsh executable.</param>
+        /// <returns>
+        /// The exit code of exec if it fails.
+        /// If exec succeeds, this process is overwritten so we never actually return.
+        /// </returns>
+        private static int ExecPwshLogin(string[] args, int loginArgIndex, string pwshPath, bool isMacOS)
+        {
+            // Create input for /bin/sh that execs pwsh
+            int quotedPwshPathLength = GetQuotedPathLength(pwshPath);
+
+            // /bin/sh does not support the exec -a feature
+            int pwshExecInvocationLength = isMacOS
+                ? quotedPwshPathLength + 19  // exec -a +pwsh '{pwshPath}' "$@"
+                : quotedPwshPathLength + 10; // exec '{pwshPath}' "$@"
```

Comment:
Talked to @rjmholt about this optimization for platform that supports `exec -a`.
If environment variable can be preserved after `exec`, then using an env variable might be better in that we wouldn't need to check for `-login` or `-l` for Linux as well.

Reply:
Will document. Also going to investigate using an environment variable.